### PR TITLE
Trac3090

### PR DESCRIPTION
**Pull request by rpetry**
Ticket #3090 - changes to handle correctly gene identifiers containing plus and colon characters in url re-written gene/<geneid> and experiment/<expacc>/<geneid> type urls.

### CODE REVIEW

File: atlas-web/src/main/webapp/WEB-INF/jsp/genepage/gene.jsp [hunk 1]
```diff
@@ -350,7 +350,7 @@ $(document).ready(function() {
                         <td align="left">
                             <c:set var="orthoIds" value=""/>
                             <c:forEach var="ortholog" items="${orthologs}">
-                                <a href="${pageContext.request.contextPath}/gene/${ortholog.geneIdentifier}"
+                                <a href="${pageContext.request.contextPath}/gene/${ortholog.urlRewriteEncodedGeneIdentifier}"
```
Comment (olgamelnichuk) [inline atlas-web/src/main/webapp/WEB-INF/jsp/genepage/gene.jsp]: Why not to use ${pageContext.request.contextPath}/gene?gid=${ortholog.geneIdentifier} here?

File: atlas-web/src/main/webapp/WEB-INF/jsp/genepage/gene.jsp [hunk 2]
```diff
@@ -384,15 +384,18 @@ $(document).ready(function() {
...
+                    <tr>
+                        <td class="geneAnnotHeader">Search EB-eye</td>
+                        <td align="left">
+                            <a title="Show gene annotation" target="_blank"
+                               href="http://www.ebi.ac.uk/ebisearch/search.ebi?db=allebi&query=${atlasGene.geneIdentifier}&requestFrom=ebi_index&submit=+FIND+">
+                                    ${atlasGene.geneIdentifier}
+                            </a>
+                        </td>
+                    </tr>
+                </c:if>
```
Comment (olgamelnichuk) [inline atlas-web/src/main/webapp/WEB-INF/jsp/genepage/gene.jsp]: Have you checked if EB-eye query="${atlasGene.geneIdentifier}"&..  works?

File: atlas-web/src/main/java/ae3/model/AtlasGene.java [hunk 3]
```diff
@@ -265,6 +266,17 @@ public class AtlasGene {
...
     /**
+     * Returns primary gene identifier, first url-encoded, then all percent characters url-encoded again
+     * Both encodings are needed to make urlrewrite rules work with gene identifiers containing e.g. '+'
+     * and ':' characters in gene/<geneid> and experiment/<expacc>/<geneid> types of urls.
+     *
+     * @return gene identifier
+     */
+    public String getUrlRewriteEncodedGeneIdentifier() {
+        return HtmlHelper.escapeURL(getGeneIdentifier()).replaceAll("\\%", "%25");
+    }
```
Comment (olgamelnichuk) [inline atlas-web/src/main/java/ae3/model/AtlasGene.java]: it is bad practice to put utility code to entity classes.  If you really need this encoding stuff then put it in the corresponding utility class or make jsp function if it is used in UI only.

File: atlas-web/src/main/webapp/WEB-INF/jsp/genepage/gene.jsp (hunk 1 above)
Comment (rpetry) [inline atlas-web/src/main/webapp/WEB-INF/jsp/genepage/gene.jsp]: As discussed, I believe the above confuse the user as to what the public API for accessing genes and experiments is. We agreed that for the point release we would fix the currently known issues with the urlrewrite functionality, but leave the interface as is. 
Also as agreed, I've opened another ticket targeted for 2.0.10, #3095, for the team as a whole to discuss rationale for using/future options for urlrewrite functionality in Atlas.

File: atlas-web/src/main/java/ae3/model/AtlasGene.java (hunk 3 above)
Comment (rpetry) [inline atlas-web/src/main/java/ae3/model/AtlasGene.java]: Agreed - moved out the functionality to HtmlHelper and made available to jsp via util.tld; also re-implemented in javascript

File: atlas-model/src/main/java/uk/ac/ebi/microarray/atlas/model/BioEntity.java [hunk 4]
```diff
@@ -35,6 +35,10 @@ public class BioEntity {
         return identifier;
     }
 
+    public void setIdentifier(String identifier) {
+        this.identifier = identifier;
+    }
+
```
Comment (olgamelnichuk) [inline atlas-model/src/main/java/uk/ac/ebi/microarray/atlas/model/BioEntity.java]: Is this change still needed?
